### PR TITLE
[DimAnalysis] Explore relations between input dimensions and fix some bugs

### PR DESCRIPTION
**Pull request by tungld**
This patch improves DimAnalysis to explore relations between input dimensions of an operation. For example, in MatMul of `A[MxN]` and `B[NxP]`, **the second dim of A** is the same as **the first dim of B**.

This patch also fixes an error in DimAnalysis in which axis in DimGroup was wrong if the dim is created by `onnx.Dim`.

For example, this is the current generated code which is wrong:
```mlir
module {
  func.func @test(%arg0: tensor<1x?xi64>) -> tensor<1x1x1x?xf32> {
    "onnx.DimGroup"(%arg0) {axis = 1 : si64, group_id = 0 : si64} : (tensor<1x?xi64>) -> ()
    %0 = onnx.Constant dense<1> : tensor<1xi64>
    %1 = "onnx.Dim"(%arg0) {axis = 1 : si64} : (tensor<1x?xi64>) -> tensor<1xi64>
    "onnx.DimGroup"(%arg0) {axis = 3 : si64, group_id = 0 : si64} : (tensor<1x?xi64>) -> ()
    %2 = "onnx.Concat"(%0, %0, %0, %1) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<4xi64>
    %3 = onnx.ConstantOfShape(%2) {value = dense<0.000000e+00> : tensor<1xf32>} : (tensor<4xi64>) -> tensor<1x1x1x?xf32>
    "onnx.DimGroup"(%3) {axis = 3 : si64, group_id = 0 : si64} : (tensor<1x1x1x?xf32>) -> ()
    return %3 : tensor<1x1x1x?xf32>
  }
}
```

`"onnx.DimGroup"(%arg0) {axis = 3 : si64, group_id = 0 : si64} : (tensor<1x?xi64>) -> ()`: `axis` must be `1` instead of `3`.

### CODE REVIEW

**Review by AlexandreEichenberger — COMMENTED** [review]
Minimally we should add all variants of `MatMul`. Added other candidate operations here for future references.

File: src/Dialect/ONNX/ONNXDimAnalysis.cpp [hunk 1]
```diff
@@ -68,21 +68,30 @@ static bool areOverlapping(
...
+    // A constant of -1 to define a dynamic value, e.g. -1 in Reshape means a
+    // dynamic dimension computed from the other dimensions. It's a contant, no
+    // need to insert it.
+    if (isa<ONNXConstantOp>(op))
+      okToInsert = false;
+    else if (auto dimOp = dyn_cast<ONNXDimOp>(op)) {
+      // The correct axis is from ONNXDimOp.
+      axis = dimOp.getAxis();
+      okToInsert = true;
+    } else if (isa<ONNXCastOp>(op) || tensorType.isDynamicDim(axis))
```
Comment (chentong319) [inline src/Dialect/ONNX/ONNXDimAnalysis.cpp:87]: I am trying the understand the difference: for DimOp, the axis is the for the input, while for other op, the axis is for the output. The outputIE of other ops is somehow connected to its inputIE by shape analysis, but not for DimOp? Special case is always questionable to me.

Comment (tungld) [reply src/Dialect/ONNX/ONNXDimAnalysis.cpp:87]: Tong, the relationship is that inputIE is defined by DimOp.

File: src/Dialect/ONNX/ONNXDimAnalysis.cpp [hunk 2]
```diff
@@ -122,9 +131,61 @@ static void findAndAddSameDim(const QuestionmarkIndexExpr &qmOuputIE,
...
-/// Use this function for operations that use adaptor to compute shape.
+/// Given a dynamic dimension of a tensor, find the same dynamic dimensions in
+/// the input tensors of the consuming operator.
+///
+/// For example, in MatMul(A, B) : MxN * NxP, dimA[1] = dimB[0] = N.
+static void exploreSameDimsFromConsumingOperators(
+    const DimAnalysis::DimT &dim, DimAnalysis::DimSetT &sameDims) {
+  LLVM_DEBUG(llvm::dbgs() << "Explore using consuming operators\n");
+  for (Operation *op : dim.first.getUsers()) {
+    if (auto matmulOp = dyn_cast<ONNXMatMulOp>(op)) {
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/Dialect/ONNX/ONNXDimAnalysis.cpp]: @AlexandreEichenberger I added support for MatMulInteger, Concat, LSTM, GRU, and RNN.

For batch norm, it is rewritten into Mul and Add: https://github.com/onnx/onnx-mlir/blob/main/src/Dialect/ONNX/Rewrite.td#L770

> Convolution also has corresponding sizes between image and filter for C, and 'imageandbias`.

What are image and filer? I didn't find them in Conv's inputs: https://github.com/onnx/onnx/blob/main/docs/Operators.md#Conv

[the remark this answers is not in the document]
Comment (AlexandreEichenberger) [reply src/Dialect/ONNX/ONNXDimAnalysis.cpp]: > What are image and filter? I didn't find them in Conv's inputs:

Sorry, use the nicknames, image is `X`, filter is `W`

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, thanks for going "overboard" with all the new cases.

**Pull request merged**